### PR TITLE
builders: create a signature builder

### PR DESCRIPTION
**Pull request by szymonlopaciuk**
Create a builder for hep.authors records. This is a continuation of [@kaplun's work](https://github.com/kaplun/inspire-schemas/blob/signatures/inspire_schemas/builders/signatures.py)

## Motivation & Context

This is a step towards inspirehep/inspire-next#2971

### CODE REVIEW

File: inspire_schemas/builders/signatures.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        self.obj['ids'] = filter(lambda uid: uid.get('schema') != schema, self.obj['ids'])
+        self._add_uid(uid, schema)
+
+    @filter_empty_parameters
+    def add_inspire_role(self, inspire_role):
+        self._ensure_list_field('inspire_roles', inspire_role)
+
+    @filter_empty_parameters
+    def add_raw_affiliation(self, raw_affiliation):
+        self._ensure_list_field('raw_affiliations', raw_affiliation)
```
Comment (michamos) [inline inspire_schemas/builders/signatures.py]: this looks wrong, as `raw_affiliations` elements are `sourced_value`s. You should do something similar to https://github.com/inspirehep/inspire-schemas/blob/ac809becb858cfe2a1c695c18f00014109138dca/inspire_schemas/builders/references.py#L187-L196

File: inspire_schemas/utils.py [hunk 2]
```diff
@@ -293,6 +295,62 @@
...
+    Args:
+        uid (string): a UID string
+
+    Returns:
+        Tuple[string, string]: a tuple (schema, uid) where:
+        - schema: a schema of the UID or None if not recognised
+        - uid: the UID normalized to comply with the id.json schema
+    """
+    if re.compile(r'^\d{8}$').match(uid):
+        # Could be either VIAF or INSPIRE ID, give up
```
Comment (michamos) [inline inspire_schemas/utils.py]: How should one add an INSPIRE ID then?

File: inspire_schemas/builders/signatures.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    def set_uid(self, uid):
+        """Set a unique ID.
+
+        If a UID of a given schema already exists in a record it will
+        be overwritten, otherwise it will be appended to the record.
+
+        Args:
+            uid (string): unique identifier
+        """
+        schema, uid = guess_author_uid_schema_and_normalize(uid)
```
Comment (michamos) [inline inspire_schemas/builders/signatures.py]: I wouldn't do the guessing if you cannot always guess reliably. I think it's better if `set_uid` takes also `schema` as argument.

File: inspire_schemas/builders/signatures.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        if not schema:
+            return
+        self._ensure_field('ids', [])
+        self.obj['ids'] = filter(lambda uid: uid.get('schema') != schema, self.obj['ids'])
```
Comment (michamos) [inline inspire_schemas/builders/signatures.py]: More than one identifier with the same schema is strange, but is in principle allowed, so you shouldn't filter here.

File: inspire_schemas/builders/signatures.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+import idutils
+
+from ..utils import (
+    filter_empty_parameters,
+    guess_author_uid_schema_and_normalize,
+)
+
+
+class SignatureBuilder(object):
+    """Class used for building JSON signatures objects given simple properties.
```
Comment (michamos) [inline inspire_schemas/builders/signatures.py]: maybe also explain that signatures are authors on literature records.

File: inspire_schemas/builders/signatures.py (hunk 3 above)
Comment (szymonlopaciuk) [reply inspire_schemas/builders/signatures.py]: Should I come back to what was there before, so trying to guess those things that can be guessed reliably (e.g. ORCID, DESY, maybe BAI) with an optional schema argument, or scrape the whole thing and force the user to explicitly state it?

File: inspire_schemas/utils.py (hunk 2 above)
Comment (szymonlopaciuk) [reply inspire_schemas/utils.py]: INSPIRE-12345678, but I see your point, maybe it's better to remove ambiguity

File: inspire_schemas/builders/signatures.py (hunk 4 above)
Comment (michamos) [reply inspire_schemas/builders/signatures.py]: I see your point. Let's keep it like this then, and we'll add the other case if the need ever arises.

File: inspire_schemas/builders/signatures.py (hunk 3 above)
Comment (michamos) [reply inspire_schemas/builders/signatures.py]: :+1: that would be better. Also most IDs never appear on papers in practice. I would say that you should support those that dojson handles https://github.com/inspirehep/inspire-dojson/blob/e5096bb6c78c7e52f65319fcfbe8e2325e9bb959/inspire_dojson/hep/rules/bd1xx.py#L46

File: inspire_schemas/builders/literature.py [hunk 6]
```diff
@@ -224,13 +203,13 @@ def add_author(self, author):
 
     @filter_empty_parameters
     def make_author(self, full_name,
-                    affiliations=None,
-                    roles=None,
-                    raw_affiliations=None,
+                    affiliations=[],
```
Comment (michamos) [inline inspire_schemas/builders/literature.py]: you should never use mutable types in default arguments (see http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments) unless you really know what you are doing. In this case, you want something iterable and immutable, so you can use the empty tuple `()` instead.

File: inspire_schemas/builders/signatures.py (hunk 3 above)
Comment (michamos) [reply inspire_schemas/builders/signatures.py]: here the guessed schema is overriding the one passed to the function. It should be the other way around.

File: inspire_schemas/utils.py [hunk 7]
```diff
@@ -293,6 +295,62 @@
...
+        if (
+                {'source', 'material'}.issuperset(my_kwargs) or not my_kwargs
+        ) and args == ():
+            return
+        return func(self, *args, **my_kwargs)
+
+    return func_wrapper
+
+
+def guess_author_uid_schema_and_normalize(uid):
```
Comment (michamos) [inline inspire_schemas/utils.py]: could you change this according to https://github.com/inspirehep/inspire-schemas/pull/276/files#r152972797 ?

File: inspire_schemas/utils.py [hunk 8]
```diff
@@ -318,38 +318,57 @@ def func_wrapper(self, *args, **kwargs):
...
+    Raise:
+        ValueError: if specified schema is not matching the given UID
     """
-    if re.compile(r'^\d{8}$').match(uid):
-        # Could be either VIAF or INSPIRE ID, give up
-        return None, uid
+    if idutils.is_orcid(uid) and schema in (None, 'ORCID'):
+        return idutils.normalize_orcid(uid), 'ORCID'
+
+    RE_MATCH = {
```
Comment (michamos) [inline inspire_schemas/utils.py]: better to have this as a global, outside of the function, so you don't rebuild this on every function call.

File: inspire_schemas/utils.py [hunk 9]
```diff
@@ -293,6 +304,81 @@
...
+        raise UnknownUIDSchema(uid)
+
+    if schema:
+        normalized_uid = _get_uid_normalized_in_schema(uid, schema)
+        if normalized_uid:
+            return normalized_uid, schema
+        else:
+            raise SchemaUIDConflict(schema, uid)
+
+    match_schema, normalized_uid = None, None
```
Comment (michamos) [inline inspire_schemas/utils.py]: you only need the first one for testing for duplicates (`matched_schema` sounds better), so no need to initialize both here.

File: inspire_schemas/utils.py [hunk 10]
```diff
@@ -293,6 +304,81 @@
...
+    for candidate_schema in _RE_AUTHORS_UID:
+        candidate_uid = _get_uid_normalized_in_schema(uid, candidate_schema)
```
Comment (michamos) [inline inspire_schemas/utils.py]: `candidate_uid` -> `normalized_uid`

File: inspire_schemas/utils.py [hunk 11]
```diff
@@ -293,6 +304,81 @@
...
+        if candidate_uid:
+            if match_schema:
```
Comment (michamos) [inline inspire_schemas/utils.py]: `if ... and ...:`

File: inspire_schemas/utils.py [hunk 12]
```diff
@@ -293,6 +304,81 @@
...
+                # Valid against more than one candidate schema, ambiguous
+                raise UnknownUIDSchema(uid)
+            match_schema = candidate_schema
+            normalized_uid = candidate_uid
```
Comment (michamos) [inline inspire_schemas/utils.py]: line not needed with previous change.

File: tests/unit/test_utils.py [hunk 13]
```diff
@@ -678,3 +678,44 @@ def test_convert_new_publication_info_to_old_handles_year_added_to_volumes():
...
+])
+def test_author_id_normalize_and_schema(
+        uid, explicit_schema, expected_uid, expected_schema):
+    normalized_uid, guessed_schema = utils.author_id_normalize_and_schema(uid, explicit_schema)
+    assert guessed_schema == expected_schema
+    assert normalized_uid == expected_uid
+
+
+def test_author_id_normalize_and_schema_unknown():
+    try:
```
Comment (michamos) [inline tests/unit/test_utils.py]: to assert that an exception has been raised, you should use the [`pytest.raises`](https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions) context manager.

File: tests/unit/test_signature_builder.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+    builder = SignatureBuilder({'existing': ['old']})
+    builder._ensure_list_field('created', 'new')
+    builder._ensure_list_field('existing', 'new')
+    result = builder.obj
+
+    assert expected == result
+
+
+def test_signature_builder(subschema):
```
Comment (michamos) [inline tests/unit/test_signature_builder.py]: you should also test the case where you initialize the builder with an existing author.

File: inspire_schemas/utils.py [hunk 15]
```diff
@@ -363,18 +363,18 @@ def _get_uid_normalized_in_schema(_uid, _schema):
         else:
             raise SchemaUIDConflict(schema, uid)
 
-    match_schema, normalized_uid = None, None
+    matched_schema = None
```
Comment (michamos) [inline inspire_schemas/utils.py]: I now see why you needed both now, it was better before, sorry.
